code/utils.py

```diff
@@ -238,10 +252,33 @@ def ez_score(true_labels, predict_labels, num_labels):
             true, 
             pred, 
             labels=[e for e in range(1, num_labels)],
-            average='micro',
+            average='weighted',
             zero_division=1,  # set score to 1 when all labels and predictions are 0
         )
-    return total/true_labels.shape[0]
+    return total/(len(true_labels) + 1e-10)
+
+
+def span_f1(gold, pred):
+    """
+    Borrowed from: https://github.com/jerbarnes/sentiment_graphs
+    """
+    tp, fp, fn = 0, 0, 0
+    for gold_labels, pred_labels in zip(gold, pred):
+        for gold_label, pred_label in zip(gold_labels, pred_labels):
+            # TP
+            if gold_label == pred_label:
+                tp += 1
+            #FP
+            if gold_label != pred_label and pred_label > 0:
+                fp += 1
+            #FN
+            if gold_label > 0 and pred_label != gold_label:
+                fn += 1
+    prec = tp / (tp + fp + 1e-6)
+    rec = tp / (tp + fn + 1e-6)
+    f1 = 2 * prec * rec / (prec + rec + 1e-6)
+    # return prec, rec, f1
+    return f1
```

Comment:
Jeremy recommended against RACL and IMN metrics, since average target scope range in English data (1-2) was much smaller than that for the Norwegian data (2-30) (double check his range limits). 

This span metric treats predictions as follows:
> - True positive: predicted label is correct
> - False positive: prediction label present, but not correct value (both `gold > 0` and `gold==0` can occur here)
> - False negative: gold label should be present, but predicted value is wrong (both `pred > 0` and `pred==0` can occur here)

Reply:
Our false negative isn't really a false negative. Should be:
- predicted label is 0, but true > 0
```
if pred_label == 0 and gold_label > 0:
    fn += 1
```

Reply:
Why is this called span when it doesn't even take into account scope? 

Checks token-wise F1 for joint tri-label.
Slightly more strict on presently labels, with wrong tag (`B` or `I`) than a regular F1 would be.
Recall basically measures only how many wrong 0 were predicted, i.e. wrongly estimated 1s or 2s won't show up here.